### PR TITLE
PR: Workaround for Qt-bug where cursor moves unintuitively after text deletion

### LINKED ISSUE #12663: Cursor moves position after deleting a character and changing line
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [x] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description

When replacing a simple string in selected text with `Replace selection`, Spyder takes a long time to replace it and to return control to the user.

### What steps reproduce the problem?

1. Open a new Python file (or use `temp.py` or any existing Python file)
2. Write three lines with `# Spyder is great` (each in its own line)
3. Move the cursor in the first line in front of `great`
4. Press `Del` to delete the character `g` from the word `great`
5. Press the key `Cursor down` once

### What is the expected output? What do you see instead?

I would expect that the cursor stays in the position before the character `g` in the next line, but it moves between `g` and `r`.

If I delete the character `g` with `BACKSPACE`, and move down one line, the cursor stays in the position between `g` and `r`, as expected.

## Versions

* Spyder version: 4.1.2
* Python version: 3.6.9. 64-bit
* Qt version: 5.12.7
* PyQt version: 5.12.3
* Operating System name/version: Linux Mint 19.3 Cinnamon

### Dependencies

```

# Mandatory:
atomicwrites >=1.2.0           :  1.3.0 (OK)
chardet >=2.0.0                :  3.0.4 (OK)
cloudpickle >=0.5.0            :  1.3.0 (OK)
diff_match_patch >=20181111    :  20181111 (OK)
intervaltree                   :  None (OK)
IPython >=4.0                  :  7.13.0 (OK)
jedi =0.15.2                   :  0.15.2 (OK)
keyring                        :  None (OK)
nbconvert >=4.0                :  5.6.1 (OK)
numpydoc >=0.6.0               :  0.9.2 (OK)
parso =0.5.2                   :  0.5.2 (OK)
pexpect >=4.4.0                :  4.8.0 (OK)
pickleshare >=0.4              :  0.7.5 (OK)
psutil >=5.3                   :  5.7.0 (OK)
pygments >=2.0                 :  2.6.1 (OK)
pylint >=0.25                  :  2.4.4 (OK)
pyls >=0.31.9;<0.32.0          :  0.31.9 (OK)
qdarkstyle >=2.8               :  2.8 (OK)
qtawesome >=0.5.7              :  0.7.0 (OK)
qtconsole >=4.6.0              :  4.7.1 (OK)
qtpy >=1.5.0                   :  1.9.0 (OK)
sphinx >=0.6.6                 :  2.4.4 (OK)
spyder_kernels >=1.9.0;<1.10.0 :  1.9.0 (OK)
watchdog                       :  None (OK)
xdg >=0.26                     :  0.26 (OK)
zmq >=17                       :  19.0.0 (OK)

# Optional:
cython >=0.21                  :  None (OK)
matplotlib >=2.0.0             :  None (OK)
numpy >=1.7                    :  1.18.2 (OK)
pandas >=0.13.1                :  None (OK)
scipy >=0.17.0                 :  1.4.1 (OK)
sympy >=0.7.3                  :  None (OK)
```

### PR DESCRIPTION
**Pull request by hengin**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

This pull-request provides workarounds for a quirk in QTextEdit that is illustrated in the GIFs below:
spyder-delete

spyder-extreme-delete

When text is deleted (whether by backspace, delete or cursor operations), the QTextEdit does not forget its last horizontal position. When navigating Up/Down right after a delete operation, the horizontal cursor will be restored to that position. Can be very detrimental to the editing experience!

The root cause is a 7 year old Qt bug: [QTBUG-35861](https://bugreports.qt.io/browse/QTBUG-35861). It seems unlikely to ever be fixed upstream. Still, I've added a regression test `test_qtbug35861` that will fail if the upstream bug is fixed.

~~Note that one edge case (deleting a selection) is only covered when the "delete" shortcut is bound to the "Delete" key. This is the default setting, but can be overridden by users.~~ Fixed as of 3cc8f3e2bcb4717.

EDIT: For some reason that edge case also fails in CI Linux tests. But I can't reproduce it on my Linux machine. I'll need some help with this. (not fixed by 3cc8f3e2bcb4717).

### Issue(s) Resolved

Fixes #12663

### Affirmation

By submitting this Pull Request and typing my username below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: hengin

### CODE REVIEW

File: spyder/plugins/editor/widgets/tests/test_codeeditor.py [hunk 1]
```diff
@@ -366,5 +366,144 @@ def test_brace_match(editorbot):
         assert editor.bracepos == expected
 
 
+def test_editor_backspace_char(editorbot):
+    """Regression test for issue spyder-ide/spyder#12663."""
+    qtbot, editor = editorbot
+    text = "0123456789\nabcdefghij\n9876543210\njihgfedcba\n"
+    editor.set_text(text)
+    expectedCol = 7
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_codeeditor.py:374]: Please rename this variable to `expected_column` because we only camel case variables to name classes.

File: spyder/plugins/editor/widgets/tests/test_codeeditor.py [hunk 2]
```diff
@@ -366,5 +366,144 @@ def test_brace_match(editorbot):
...
+        qtbot.keyPress(editor, Qt.Key_Up)
+        assert editor.textCursor().columnNumber() == expectedCol
+
+
+def test_editor_backspace_selection(editorbot):
+    """Regression test for issue spyder-ide/spyder#12663."""
+    qtbot, editor = editorbot
+    text = "0123456789\nabcdefghij\n9876543210\njihgfedcba\n"
+    editor.set_text(text)
+    expectedCol = 5
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_codeeditor.py:398]: expectedCol -> expected_column

File: spyder/plugins/editor/widgets/tests/test_codeeditor.py [hunk 3]
```diff
@@ -366,5 +366,144 @@ def test_brace_match(editorbot):
...
+    qtbot.keyPress(editor, Qt.Key_Down)
+    assert editor.textCursor().columnNumber() == expectedCol
+
+
+def test_editor_delete_char(editorbot):
+    """Regression test for issue spyder-ide/spyder#12663."""
+    qtbot, editor = editorbot
+    text = "0123456789\nabcdefghij\n9876543210\njihgfedcba\n"
+    editor.set_text(text)
+    expectedCol = 2
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_codeeditor.py:426]: expectedCol -> expected_column

File: spyder/plugins/editor/widgets/tests/test_codeeditor.py [hunk 4]
```diff
@@ -366,5 +366,144 @@ def test_brace_match(editorbot):
...
+        qtbot.keyPress(editor, Qt.Key_Up)
+        assert editor.textCursor().columnNumber() == expectedCol
+
+
+def test_editor_delete_selection(editorbot):
+    """Regression test for issue spyder-ide/spyder#12663."""
+    qtbot, editor = editorbot
+    text = "0123456789\nabcdefghij\n9876543210\njihgfedcba\n"
+    editor.set_text(text)
+    expectedCol = 5
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_codeeditor.py:448]: expectedCol -> expected_column

File: spyder/plugins/editor/widgets/tests/test_codeeditor.py [hunk 5]
```diff
@@ -366,5 +366,144 @@ def test_brace_match(editorbot):
...
+    # This first subtest does not trigger the original bug
+    for press in range(3):
+        qtbot.keyPress(editor, Qt.Key_Left, Qt.ShiftModifier)
+    expectedCol -= 3
+    qtbot.keyPress(editor, Qt.Key_Delete)
+    assert editor.textCursor().columnNumber() == expectedCol
+    qtbot.keyPress(editor, Qt.Key_Down)
+    assert editor.textCursor().columnNumber() == expectedCol
+
+    # However, this second subtest does trigger the original bug
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_codeeditor.py:462]: This means that this part will catch a regression in case we change things and they don't stop working as they are in this PR?

File: spyder/plugins/editor/widgets/tests/test_codeeditor.py [hunk 6]
```diff
@@ -366,5 +366,144 @@ def test_brace_match(editorbot):
...
+    qtbot.addWidget(widget)
+    widget.show()
+
+    cursor = widget.textCursor()
+    cursor.setPosition(0)
+    # Build the text from a single character since a non-fixed width
+    # font is used by default.
+    cursor.insertText("0000000000\n"*5)
+
+    expectedCol = 5
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_codeeditor.py:490]: expectedCol -> expected_column

**Review by ccordoba12 — COMMENTED** [review]
Thanks @hengin for your work on this! It looks pretty good to me, except for some minor formatting shortcomings.

About the failures on Linux, you can skip your new tests on that platform if you're certain they're passing locally. The thing is our tests run in servers with no real display (it's emulated by a frame buffer Xorg server instead), so sometimes is tricky to make them pass in those circumstances.

In order to do that, you need to add the following decorator to them:

    @pytest.mark.skipif(sys.platform.startswith('linux'), reason='Fail on Linux')

And (of course), import the `sys` module on top of that file.

**Comment by juanis2112** [comment]
I also tested this and it works fine now. Thank you very much @hengin for fixing this!

File: spyder/plugins/editor/widgets/tests/test_codeeditor.py (hunk 5 above)
Comment (hengin) [reply spyder/plugins/editor/widgets/tests/test_codeeditor.py:462]: The second subtest (last two asserts) would catch a regression, yes. In the sense that the test will fail in 4.2.1, but succeds with things as they are in this PR.

**Comment by hengin** [comment]
Thanks for testing @juanis2112! That gives me more confidence that the fix will at least work most of the time.

**Review by ccordoba12 — APPROVED** [review]
Thanks @hengin for your help with these complex bugs! We really value it.

Also, thanks @juanis2112 for manually testing this one (I didn't have time to do it myself).

**Comment by ccordoba12** [comment]
@hengin, a small suggestion for next time: After running `git checkout 4.x` (or `git checkout master`, in case you want to solve a bug for Spyder 5), please run `git checkout -b fix-isssue-xxxx`. Or, instead of `fix-isssue-xxxx`, please select another branch name that makes sense for you.

That will create a branch for your work, which will allow us to help you in case we need to add commits to your branch or solve difficult merges or rebases for you (if necessary). That's not possible if you commit directly to either the `4.x` or `master` branches in your fork.
